cpg_pipes/pipeline/pipeline.py

```diff
@@ -0,0 +1,960 @@
+"""
+Provides `Pipeline` class and a `@stage` decorator that allow to define pipeline
+stages and plug them together.
+
+A Stage describes the job that are added to Hail Batch, and outputs that are expected
+to be produced. Each stage acts on a `Target`, which can be of a different level:
+a `Sample`, a `Dataset`, a `Cohort` (= all input datasets combined). Pipeline plugs
+stages together by resolving dependencies between different levels accordingly.
+
+Examples of pipelines can be found in the `pipelines/` folder in the repository root.
+"""
+import functools
+import logging
+import pathlib
+import shutil
+import tempfile
+import time
+from abc import ABC, abstractmethod
+from typing import Callable, cast, Union, TypeVar, Generic, Any, Optional, Type
+
+from cloudpathlib import CloudPath
+import hailtop.batch as hb
+from hailtop.batch.job import Job
+
+from .exceptions import PipelineError
+from .targets import Target, Dataset, Sample, Cohort
+from .. import Path, to_path
+from ..providers import (
+    Cloud,
+    Namespace,
+    StoragePolicy,
+    StatusReporterType,
+    InputProviderType,
+)
+from ..providers.status import StatusReporter
+from ..providers.cpg import (
+    CpgStorageProvider,
+    SmdbStatusReporter,
+    SmdbInputsProvider,
+    SMDB,
+)
+from ..refdata import RefData
+from ..utils import exists
+from ..hb.batch import Batch, setup_batch
+from ..hb.prev_job import PrevJob
+
+logger = logging.getLogger(__file__)
+
+
+StageDecorator = Callable[..., 'Stage']
+
+
+# We record each initialised Stage subclass, so we know the default stage
+# list for the case when the user doesn't pass them explicitly with set_stages()
+_ALL_DEFINED_STAGES = []
+
+
+def stage(
```

Comment:
The order of declarations in this File is a bit strange, to me it would be something like:

- StageInput / StageOutput
- Stage
- stage decorator
- Pipeline